File: DSA/Animesh_jha_590016813/Day_27_question_3.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* removeKthNode(struct Node* head, int k) {
    if (head == NULL || k <= 0)
        return head;

    if (k == 1) {
        while (head != NULL) {
            struct Node* temp = head;
            head = head->next;
            free(temp);
        }
        return NULL;
    }

    struct Node* current = head;
    struct Node* prev = NULL;
    int count = 1;

    while (current != NULL) {
        if (count % k == 0) {
            prev->next = current->next;
            free(current);
            current = prev->next;
        } else {
            prev = current;
            current = current->next;
        }
        count++;
    }
    return head;
}

struct Node* insertEnd(struct Node* head, int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;

    if (head == NULL)
        return newNode;

    struct Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;

    temp->next = newNode;
    return head;
}

void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d", head->data);
        if (head->next)
            printf(" -> ");
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = NULL;
    for (int i = 1; i <= 8; i++) {
        head = insertEnd(head, i);
    }

    int k = 2;

    printf("Original list:\n");
    printList(head);

    head = removeKthNode(head, k);

    printf("List after removing every %dth node:\n", k);
    printList(head);

    return 0;
}
 

File: DSA/LakshaKundra_590013237/Day_27_Question_3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def removeKthNode(head, k):
    if k == 0:
        return head
    if k == 1:
        return None
    curr = head
    count = 1
    while curr:
        for _ in range(k - 2):
            if curr.next:
                curr = curr.next
            else:
                return head
        if curr.next:
            curr.next = curr.next.next
            curr = curr.next
        else:
            break
    return head

def printList(head):
    while head:
        print(head.data, end=" ")
        head = head.next
    print()

def createList(arr):
    if not arr:
        return None
    head = Node(arr[0])
    curr = head
    for val in arr[1:]:
        curr.next = Node(val)
        curr = curr.next
    return head

head = createList([1, 2, 3, 4, 5, 6, 7, 8])
head = removeKthNode(head, 2)
printList(head)

head = createList([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
head = removeKthNode(head, 3)
printList(head)
